### ISSUE TITLE
Picocom, failed to add port.

### ISSUE THREAD

**Issue by sittim** [npat-efault/picocom#94]
On Arch OS, picocom ver3.1, I get the following error:

`FATAL: failed to add port: Cannot get the device attributes: Inappropriate ioctl for device`

Using Linux Arch OS, help output:

```
[I] /mnt/c/Users/timofeysitnikov/HydroGuardFW/hw_1_5/Production~> picocom --help
picocom v3.1

Compiled-in options:
  TTY_Q_SZ is 0
  HIGH_BAUD is enabled
  USE_FLOCK is enabled
  LINENOISE is enabled
  HISTFILE is: .picocom_history
  USE_CUSTOM_BAUD is enabled
```

Attempt to open picocom:

```
[N] /mnt/c/Users/timofeysitnikov/HydroGuardFW/hw_1_5/Production~> picocom -b 19200 /dev/ttyS17
picocom v3.1

port is        : /dev/ttyS17
flowcontrol    : none
baudrate is    : 19200
parity is      : none
databits are   : 8
stopbits are   : 1
escape is      : C-a
local echo is  : no
noinit is      : no
noreset is     : no
hangup is      : no
nolock is      : no
send_cmd is    : sz -vv
receive_cmd is : rz -vv -E
imap is        :
omap is        :
emap is        : crcrlf,delbs,
logfile is     : none
initstring     : none
exit_after is  : not set
exit is        : no


FATAL: failed to add port: Cannot get the device attributes: Inappropriate ioctl for device
```

On Ubuntu 16.04, picocom v2.3a, works fine, no issues.

**Comment by sittim** [comment]
Release 3.0 works fine

**Comment by npat-efault** [comment]
Try to compile release 3.1 without custom baud-rate support. Something like this:

    CPPFLAGS=-DNO_CUSTOM_BAUD make clean
    CPPFLAGS=-DNO_CUSTOM_BAUD make

Then run picocom like:

    ./picocom /dev/ttyS17

If this works, then compile again *with* custom baudrate support:

    make clean
    make

and try to open some dummy pseudo-ttty:

    open a terminal window
    $ tty
    /dev/pts/8
    (form another window)
    $ ./picocom /dev/pts/8

See if it works, and report back

**Comment by npat-efault** [comment]
Also: I downloaded the picocom 3.1-1 binary package, from here

  https://www.archlinux.org/packages/community/x86_64/picocom/

I directly ran the picocom binary found in it (on my Ubuntu 16.04 system). It works fine with my serial ports.

What type of serial port is this `/dev/ttyS17`? What make is it (some kind of multiport card, or something), and what driver does it use?

In any case, I believe that if you recompile w/o custom baudrate support, it should work...

**Comment by npat-efault** [comment]
I'm guessing that for some reason (and *I don't understand why*) the specific serial-port you are using is not accepting the new "`termios2`" ioctl's (`TCGETS2`, `TCSETSF2`, etc) that are used when picocom is compiled with custom baudrate support (which happens by default starting with release 3.1).

If my assumption is correct then [this commit, may help](ee23af0cb96ed3bd226f92dad663bc9bba865b2f). 

Clone or download picocom tip, compile as usual (this will enable custom baudrate support), and then run picocom. If it doesn't work as before, then define the `NO_CUSTOM_BAUD` environment variable, and try again. Something like this:

    NO_CUSTOM_BAUD=1 ./picocom ... 

This will make picocom use the "old" ioctls, as if it was compiled without custom-baudrate support.

**Comment by sittim** [comment]
This is on Windows WSL, so its the Linux interface on Windows.  Its just a regular com to this device: http://www.ti.com/tool/MSP-FET , it has a USB to serial.  I am swamped, but I will try your recommendations above shortly.

**Comment by npat-efault** [comment]
Ah! I see!! Then almost certainly that the "new" ioctl's are the problem in this case. Doing a very quick  google, I found [this](https://github.com/Microsoft/WSL/issues/2595), which certainly explains it.

So, you can *either* compile your picocom w/o custom baudrate support, *or* get picocom tip, compile it with custom baudrate support (default), and disable it at run-time using the environment variable, as described in the previous post...

Actually, If and whenever you find some time, please **do try** the second approach (picocom tip, compiled with custom baud support, disable custom baudrate support by defining the environment variable). And when you do, drop a line if it worked. This will be a good test to the new feature...

**Comment by sittim** [comment]
I will do that by the end of the week.

**Comment by shanselman** [comment]
@sittim did this happen elsewhere?

**Comment by sittim** [comment]
No, it was just wsl.

**Comment by imliubo** [comment]
Has this problem been solved? I also encountered the same error.

WSL ubuntu 20.04
picocom v3.2a

**Comment by nakotami** [comment]
I have the same problem with WSL Debian. Any updates?

**Comment by SunJun8** [comment]
Ubuntu 20.04 WSL, picocom v3.1, encountered the same problem

**Comment by rohit-kumar-j** [comment]
Still not working with WSL Ubuntu 20.04, picocom V3.1.
Nor is it working with picocom 3.0.. 😓

**Comment by NewComer00** [comment]
Same problem on Cygwin. But if I first establish the serial connection using another program like screen, then quit to use picocom, the error will be gone and now I'm able to connect. Maybe some problems during initialization?

picocom v3.2a
Windows 10 21H2 19044
Cygwin DLL 3.3.6-341.x86_64

**Comment by feilipu** [comment]
Just to add that I've recently seen the same problem using Ubuntu 22.04 on WSL1 using the current Release 3.1.
Release 3.1 is unusable on WSL1.

I resolved it by installing the [Ubuntu 18.04 package](https://packages.ubuntu.com/bionic/comm/picocom) which is Release 2.2.
